instrumentation/kafka-clients/kafka-clients-0.11/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/KafkaTracing.java

```diff
@@ -0,0 +1,125 @@
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.kafka.KafkaHeadersGetter;
+import io.opentelemetry.instrumentation.kafka.KafkaHeadersSetter;
+import io.opentelemetry.instrumentation.kafka.KafkaSingletons;
+import io.opentelemetry.instrumentation.kafka.ReceivedRecords;
+import io.opentelemetry.instrumentation.kafka.Timer;
+import java.util.Objects;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.header.Headers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaTracing {
+
+  private static final Logger logger = LoggerFactory.getLogger(KafkaTracing.class);
+
+  private static final TextMapGetter<Headers> GETTER = new KafkaHeadersGetter();
+
+  private static final TextMapSetter<Headers> SETTER = new KafkaHeadersSetter();
+
+  private Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter =
+      KafkaSingletons.producerInstrumenter();
+  private Instrumenter<ReceivedRecords, Void> consumerReceiveInstrumenter =
+      KafkaSingletons.consumerReceiveInstrumenter();
+  private Instrumenter<ConsumerRecord<?, ?>, Void> consumerProcessInstrumenter =
+      KafkaSingletons.consumerProcessInstrumenter();
+
+  KafkaTracing() {}
+
+  public static KafkaTracingBuilder newBuilder(OpenTelemetry openTelemetry) {
+    return new KafkaTracingBuilder(openTelemetry);
+  }
+
+  private static TextMapPropagator propagator() {
+    return GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+  }
+
+  void setProducerInstrumenter(Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter) {
+    this.producerInstrumenter = Objects.requireNonNull(producerInstrumenter);
+  }
+
+  void setConsumerReceiveInstrumenter(Instrumenter<ReceivedRecords, Void> consumerInstrumenter) {
+    this.consumerReceiveInstrumenter = Objects.requireNonNull(consumerInstrumenter);
+  }
+
+  void setConsumerProcessInstrumenter(
+      Instrumenter<ConsumerRecord<?, ?>, Void> consumerInstrumenter) {
+    this.consumerProcessInstrumenter = Objects.requireNonNull(consumerInstrumenter);
+  }
+
+  /**
+   * Build and inject span into record. Return Runnable handle to end the current span.
+   *
+   * @param record the producer record to inject span info.
+   * @return runnable to close the current span
+   */
+  public <K, V> Runnable buildAndInjectSpan(ProducerRecord<K, V> record) {
+    if (!producerInstrumenter.shouldStart(Context.current(), record)) {
```

Comment:
Please extract `Context.current()` to a local `parentContext` variable to avoid overhead caused by thread local lookup.